src/lib/gossip_net/gossip_net.ml

```diff
@@ -402,21 +405,63 @@ module Make (Message : Message_intf) : S with type msg := Message.msg = struct
 
   let create (config : Config.t)
       (implementation_list : Host_and_port.t Rpc.Implementation.t list) =
+    let t_for_restarting = ref None in
     trace_task "gossip net" (fun () ->
+        let fail m =
+          failwith
+            (Printf.sprintf "Failed to connect to Kademlia process: %s\n" m)
+        in
+        let restart_counter = ref 0 in
+        let rec handle_exn e =
+          incr restart_counter ;
+          if !restart_counter > 5 then
+            failwithf
+              "Already restarted Kademlia subprocess 5 times, dying with \
+               exception %s"
+              (Exn.to_string e) () ;
+          match Monitor.extract_exn e with
+          | Kademlia.Membership.Child_died ->
+              let t = Option.value_exn !t_for_restarting in
```

Comment:
Will this not fail always? you'd get `Some t` only after `membership` is determined, no?

Reply:
It could fail if there's an exception before `membership` is resolved. This `handle_exn` function will only get called by `try_with` after the `Membership.connect` completes. So the process would have to start successfully but fail very quickly.